### PR TITLE
8290834: Improve potentially confusing documentation on collection of profiling information

### PR DESCRIPTION
**Pull request by TheShermanTanker**
Documentation on the MethodData object incorrectly states that it is used when profiling in tiers 0 and 1, when it only does so for tier 0 (Interpreter) and tier 3, while tier 1 (Fully optimizing C1) does not collect any profile data at all. Additionally, the description for the different execution tiers is slightly confusing. This cleanup attempts to slightly better clarify how profiling is tied together between the Interpreter and C1, explain what MDO is an abbreviation for (MethodData object), and corrects the documentation for MethodData as well.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8290834](https://bugs.openjdk.org/browse/JDK-8290834): Improve potentially confusing documentation on collection of profiling information

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9598/head:pull/9598` \
`$ git checkout pull/9598`

Update a local copy of the PR: \
`$ git checkout pull/9598` \
`$ git pull https://git.openjdk.org/jdk pull/9598/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9598`

View PR using the GUI difftool: \
`$ git pr show -t 9598`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9598.diff">https://git.openjdk.org/jdk/pull/9598.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/compiler/compilationPolicy.hpp [hunk 1]
```diff
@@ -34,12 +34,18 @@ class CompileTask;
...
- *  * level 3 - C1 with full profiling (level 2 + MDO)
+ *  * level 3 - C1 with full profiling (level 2 + All data available from an MDO)
  *  * level 4 - C2
  *
+ * Note: The MethodData object is specific to the interpreter, when execution progresses to
+ * levels 2 and 3, profile data collection utilizes the separate ciMethodData instead. If any
+ * profiling information was collected from the interpreter (The method has a non-null MDO),
+ * all data from the MDO will be loaded into the ciMethodData when it is first created.
+ * (See ciMethod::method_data() in ciMethod.cpp for more details)
+ *
```
Comment (dean-long) [inline src/hotspot/share/compiler/compilationPolicy.hpp:48]: The ciMethodData is just a temporary snapshot.  Updates to the profiling data is still done through the MethodData.

Comment (TheShermanTanker) [reply src/hotspot/share/compiler/compilationPolicy.hpp:48]: The only place I could find a MethodData being created is in the method
```
// Build a MethodData* object to hold information about this method
// collected in the interpreter.
void Method::build_interpreter_method_data(const methodHandle& method, TRAPS)
```
which presumably constructs a MethodData for profiling in the Interpreter. Is there a different area where it's created (when profiling with C1) that I missed?

**Comment by dean-long** [comment]
C1 uses ciMethod::ensure_method_data(), which calls Method::build_interpreter_method_data(), to create an MDO if one wasn't already created by the interpreter.  So the name build_interpreter_method_data() is a bit misleading, because C1 will use the same MDO as the interpreter.

I also found a comment in c1_globals.hpp about C1UpdateMethodData that mentions tier1.  I think the  comment should be changed to say tier3.

File: src/hotspot/share/compiler/compilationPolicy.hpp (hunk 1 above)
Comment (TheShermanTanker) [reply src/hotspot/share/compiler/compilationPolicy.hpp]: hopefully resolved by the newest changes

File: src/hotspot/share/compiler/compilationPolicy.hpp [hunk 2]
```diff
@@ -34,11 +34,21 @@ class CompileTask;
...
- *  * level 3 - C1 with full profiling (level 2 + MDO)
- *  * level 4 - C2
+ *  * level 3 - C1 with full profiling (level 2 + All other MDO profiling information)
+ *  * level 4 - C2 with full profile guided optimization
+ *
+ * The MethodData object is used by both the interpreter and C1 to store any profiling
+ * information collected on a method (CompilationPolicy::create_mdo, compilationPolicy.cpp
+ * for the interpreter and ciMethod::ensure_method_data, ciMethod.cpp for C1), and interacts
+ * with C1 and C2 via the compiler interface. It is updated periodically as more profiling
+ * information is gathered, directly in the case of the interpreter and through ciMethodData
```
Comment (dean-long) [inline src/hotspot/share/compiler/compilationPolicy.hpp:47]: This is still a bit misleading.  The information flow between MethodData and ciMethodData is one-way.  Only MethodData are updated by the interpreter or generated code.  ciMethodData is just a read-only snapshot used during compilation.

**Comment by vnkozlov** [comment]
Please note. CI (compiler interface) and its API (ciMethod*) is used by JIT compilers C1 and C2 only **during** compilation.
Compilers (C1 and C2) can create MDO (through CI) if it is missed but they don't update data in it.
Only Interpreter and tier 3 (profiling) **compiled code** produced by C1 updates MDO. It does it directly without CI.

**Comment by vnkozlov** [comment]
@dean-long posted his comment before I finished my :)
But we are saying the same thing.

**Comment by TheShermanTanker** [comment]
> Please note. CI (compiler interface) and its API (ciMethod*) is used by JIT compilers C1 and C2 only **during** compilation. Compilers (C1 and C2) can create MDO (through CI) if it is missed but they don't update data in it. Only Interpreter and tier 3 (profiling) **compiled code** produced by C1 updates MDO. It does it directly without CI.

Ah, my mistake. I'll fix the issue asap

File: src/hotspot/share/compiler/compilationPolicy.hpp (hunk 2 above)
Comment (TheShermanTanker) [reply src/hotspot/share/compiler/compilationPolicy.hpp]: Revised

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by TheShermanTanker** [comment]
/integrate
Thanks all for the reviews!

**Review by dean-long — APPROVED** [review]

**Pull request closed without merging**